### PR TITLE
[LTC] Add _log_softmax TorchScript lowering

### LINKED ISSUE #62431: Support more ops in LazyTensor to TorchScript lowering
## 🚀 Feature
Support more ops in LazyTensor to TorchScript lowering

## Motivation

Currently Lazy Tensor lowers [a limited set of ops](https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/ts_native_functions.yaml ) to TorchScript IR. For those are not supported yet, Lazy Tensor will fall back to the Eager execution. From the correctness perspective, this works fine. But various backends optimizers would rely on Lazy Tensor to capture larger IR graph for more aggressive optimizations. 

Here is a maybe-growing list for missing ops. If you want to sign up for adding an op, simply put your name behind the item and cross it out after you are done.

- [x] addcdiv (@desertfire)
- [x] fill_ (@wconstab)
- [x] _log_softmax (@alanwaketan)
- [x] _log_softmax_backward_data (@alanwaketan)
- [x] mean (@Gamrix)
- [x] nll_loss_backward (@alanwaketan)
- [x] nll_loss_forward (@alanwaketan)
- [x] random_ (@alanwaketan)
- [x] sqrt (@kshitij12345)
- [x] zero_ (@alanwaketan)
- [x] threshold_backward (@alanwaketan)
- [x] threshold (@anijain2305)

How to add an op (using `relu` as an example, credit to @asuhan):
- https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/QUICKSTART.md - instructions for building Lazy Tensor Core

- https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/ts_native_functions.yaml - add relu
note- op name should match one found in pytorch's `native_functions.yaml` 
- https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp - add an implementation using LazyTensor::relu

- https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp - add lowering to TorchScript

- To test, `test/cpp/build/test_ptltc --gtest_filter=AtenLtcTsTensorTest.TestRelu`

A complete lowering guide can be found (credit to @asuhan):
- https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/OP_LOWERING_GUIDE.md

cc @gmagogsfm

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This patch added _log_softmax to TorchScript lowering.

Test Plan:
[.../pytorch/lazy_tensor_core] test/cpp/build/test_ptltc --gtest_filter=AtenLtcTsTensorTest.TestLogSoftmax*

Fixes #62431.

### CODE REVIEW

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp [hunk 1]
```diff
@@ -72,6 +72,13 @@ at::Tensor subtensor(const at::Tensor& tensor, int dim, int groups, int g) {
 
 }  // namespace
 
+at::Tensor LazyNativeFunctions::_log_softmax(const at::Tensor& self, int64_t dim,
+                                         bool /* half_to_float */) {
```
Comment (alanwaketan) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp:76]: @asuhan I have followed the _softmax method to omit the half_to_float parameter, but not exactly sure why it can be omitted. Will appreciate it if you can illustrate a little bit.

**Comment by alanwaketan** [comment]
Linter issues don't seem to be relate to this PR. Have filed #63109 to fix those general issues.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp (hunk 1 above)
Comment (wconstab) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp:76]: good question.  Half was not supported in LTC previously, but @desertfire just added it.  Maybe softmax needs to be updated too and both should support it.  

I guess for XLA it was not used, but it may be important for nvfuser which is obviosuly cuda-based.  

A quick search shows it is used basically for this special case of type-promotion in cuda, otherwise it is false:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/SoftMax.cpp#L29

Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp:76]: Filed #63113 to keep track of adding the support of the parameters since it might be better to do it for both operators and therefore is better as a follow up. Thanks for the illustration, @wconstab.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp [hunk 2]
```diff
@@ -265,6 +266,10 @@ class TSNodeLowering : public NodeLowering {
       return LowerLeakyReluBackward(ir::NodeCast<ir::ops::LeakyReluBackward>(
           node, ir::OpKind(at::aten::leaky_relu_backward)));
     }
+    if (node->op().op == at::aten::log_softmax) {
```
Comment (alanwaketan) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:269]: Wondering under what situation this file needs to be modified given not all operators have their lowering here. @asuhan

Comment (asuhan) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:269]: @alanwaketan Operators which are "pure" simply work by falling through here: https://github.com/pytorch/pytorch/blob/771a5ae8af25bf6b2689e55bc1fab4bcdccb7732/lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp#L309

By "pure" I mean they only have operands and don't have additional parameters. For example, `log_softmax` has an axis, convolutions have strides, dilation etc.

An accurate view of what's lowered is here: https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/ts_native_functions.yaml. In `TSNodeLowering` we had already selected the intended subset of operators.

Note that if something falls through unintentionally (we have a bug in the higher layers), we'll most likely get an error from TorchScript because the call we emit will lack those additional parameters, assuming it even passes shape inference. It's a fall-through, but it's pretty safe.

Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:269]: Thanks for the explanations. They all make perfect sense.

File: lazy_tensor_core/test/cpp/test_aten_ltc_ts_tensor.cpp [hunk 3]
```diff
@@ -6163,6 +6163,22 @@ TEST_F(AtenLtcTsTensorTest, TestLogSoftmaxCast) {
   });
 }
 
+TEST_F(AtenLtcTsTensorTest, TestLogSoftmaxWrapper) {
```
Comment (asuhan) [inline lazy_tensor_core/test/cpp/test_aten_ltc_ts_tensor.cpp:6166]: We already have a test, `TestLogSoftmax`. Why do we care about testing the `_log_softmax` directly?

**Review by asuhan — APPROVED** [review]
I'm not sure the additional test is necessary, other than that looks good.

File: lazy_tensor_core/test/cpp/test_aten_ltc_ts_tensor.cpp (hunk 3 above)
Comment (alanwaketan) [reply lazy_tensor_core/test/cpp/test_aten_ltc_ts_tensor.cpp:6166]: Just mimic what the TestSoftmaxWrapper does.

**Comment by alanwaketan** [comment]
Thanks @asuhan for approving this patch.
